src/tuner/nccl_ofi_model.c

```diff
@@ -17,90 +17,148 @@ float nccl_ofi_tuner_compute_base_cost(ncclFunc_t func, int algo, int proto)
-				  int algo,
-				  int proto,
-				  int pipe_ops,
-				  size_t size)
+static float get_base_latency(const nccl_ofi_tuner_cost_table_t *base_costs,
+			      const ncclFunc_t func,
+			      const int algo,
+			      const int proto)
+{
+	return (*base_costs)[func][algo][proto];
+}
+
+static float get_network_latency(struct nccl_ofi_tuner_model_params const *params,
+				 const int proto,
+				 float completion_overhead)
 {
-	float cost = -1;
-	float latency = 0;
-	float bw = 0;
-	int num_steps = 0;
-	int num_internode_steps = 0;
+	return proto == NCCL_PROTO_SIMPLE ? params->net_lat + completion_overhead : params->net_lat;
+}
+
+static float get_nvlink_latency(const int algo, const int proto)
+{
+	return nccl_nvlink_lat[algo][proto];
+}
+
+float nccl_ofi_tuner_scale_bandwidth(const float bandwidth, const int proto)
+{
+	/* 8B total with 4B data and 4B flags, so take a 50% hit */
+	if (proto == NCCL_PROTO_LL) {
+		return bandwidth * 0.5;
+	}
 
+	/* 120B data and 8B flags */
+	if (proto == NCCL_PROTO_LL128) {
+		return bandwidth * 0.9375;
+	}
+
+	/* No overhead otherwise */
+	return bandwidth;
+}
+
+float nccl_ofi_tuner_compute_bandwidth(struct nccl_ofi_tuner_model_params const *params,
+				       const ncclFunc_t func,
+				       const int algo)
+{
+	switch (func) {
+	case ncclFuncAllReduce:
+		switch (algo) {
+		case NCCL_ALGO_RING:
+			return params->internode_bw * params->num_rails *
+			       ofi_nccl_tuner_num_channels() / 2.0;
+		case NCCL_ALGO_NVLS_TREE:
+			return NCCL_OFI_MIN(params->intranode_bw,
+					    (params->internode_bw * params->num_rails) / 2.0) *
+			       ofi_nccl_tuner_num_channels();
+		case NCCL_ALGO_TREE:
+			return params->internode_bw * params->num_rails *
+			       ofi_nccl_tuner_num_channels() / 2.0;
+		default:
+			NCCL_OFI_TRACE(NCCL_TUNING,
+				       "Unsupported algo %d, fallback to NCCL's selection.",
+				       algo);
+			return -1;
+		}
+
+	default:
+		NCCL_OFI_TRACE(NCCL_TUNING,
+			       "Unsupported collective %d, fallback to NCCL's selection.",
+			       func);
+		return -1;
+	}
+}
+
+float nccl_ofi_tuner_compute_latency(struct nccl_ofi_tuner_model_dims const *dims,
+				     const float base_latency,
+				     const float net_latency,
+				     const float p2p_latency,
+				     const ncclFunc_t func,
+				     const int algo,
+				     const int pipe_ops)
+{
 	/*
 	 * There is more involved than the NET_COMP_OVERHEAD itself for the
 	 * simple protocol, including overheads from libfabric and NCCL's proxy
 	 * thread itself in processing a completion handed to the host by the
 	 * device. Costs associated with out-of-order completions that could
 	 * stall the pipeline should be captured here as well.
 	 */
-	const float base_latency = (*base_costs)[func][algo][proto];
-	const float net_lat = (proto == NCCL_PROTO_SIMPLE)
-				      ? params->net_lat + ofi_nccl_tuner_net_comp_overhead()
-				      : params->net_lat;
-	const float p2p_lat = nccl_nvlink_lat[algo][proto];
```

Comment:
The comment needs to move over to where this cost gets applied as well.

Reply:
lost context with updates, let me know if this is still a problem.